### PR TITLE
Reduce GPU - kernel frontend

### PR DESCRIPTION
**Pull request by mzient**
#### Why we need this PR?
*Pick one, remove the rest*
- It adds new feature needed because of we want fast, selectable axis reductions.

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     * CRTP base class
     * Stage setup and switch by stage kind
 - Affected modules and functionalities:
     * kernels/reduce module
 - Key points relevant for the review:
     * design / distribution of content across files
 - Validation and testing:
     * Unit tests
 - Documentation (including examples):
     * Doxygen, comments

**JIRA TASK**: DALI-1250

### CODE REVIEW

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+      DALI_ENFORCE(sample_shape[a] == first_sample_shape[a], make_string(
+        "Reduce: batch reduction requires that all samples have the same extent in non-reduced "
+        "dimensions.\nError at sample ", i, " axis ", a, ": the extent is ", sample_shape[a],
+        " != ", first_sample_shape[a], " in the first sample in the batch."));
+    }
+  }
+}
+
+
+inline void CheckAxes(span<const int> axes, int dim) {
```
Comment (szalpal) [inline dali/kernels/reduce/reduce_setup_utils.h:132]:
```suggestion
inline void CheckAxes(span<const int> axes, int ndims) {
```
`dim` IMHO suggests, it's an index of some specific dimension

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  uint64_t mask = 0;
+  for (auto a : axes) {
+    if (a < 0 || a >= dim)
+      throw std::out_of_range(make_string("Axis index out of range: ", a, " not in 0..", dim-1));
+    uint64_t amask = 1ul << a;
+    if (mask & amask)
+      throw std::invalid_argument(make_string("Duplicate axis index ", a));
```
Comment (szalpal) [inline dali/kernels/reduce/reduce_setup_utils.h:139]: This will only work, when `ndims<64`, right? Maybe `assert(ndims < 64)` somewhere at the beginning would be good, to document this?

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+    mask |= amask;
+  }
+}
```
Comment (szalpal) [inline dali/kernels/reduce/reduce_setup_utils.h:142]: I'm not opposed to throwing otherwise than `DALI_ENFORCE`, but since it's our current workflow, I just want to put extra attention by asking the question. Could it have negative side effects? If not, then I guess we can agree that it's ok to throw not only in `DALI_ENFORCE`

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+
+inline void CheckBatchReduce(const TensorListShape<> &tls, span<const int> axes) {
+  if (tls.num_samples() == 0)
+    return;
+
+  uint64_t mask = to_bit_mask(axes);
+  SmallVector<int, 6> non_reduced;
```
Comment (szalpal) [inline dali/kernels/reduce/reduce_setup_utils.h:112]: Why the `6` here?

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  return mask;
+}
+
+/**
+ * @param out_axes a collection of integers, supporting `clear` and `push_back`.
+ * @param dim_groups a collection of integer pairs, supporting `clear` and `push_back`
+ */
+template <typename Axes, typename DimGroups, int ndim>
+void SimplifyReduction(Axes &out_axes, DimGroups &dim_groups,
+                       const TensorListShape<ndim> &in_shape, span<const int> axes) {
```
Comment (szalpal) [inline dali/kernels/reduce/reduce_setup_utils.h:56]: It's not clear, from the documentation, what exactly is a `dim_group`. Could you put a description? (especially, because it's an output parameter)

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+template <typename ArrayLike, int ndim>
+inline void CalculateReductionFactors(ArrayLike &out, const TensorListShape<ndim> &in_shape,
+                                      span<const int> axes) {
+  for (int i = 0; i < in_shape.num_samples(); i++) {
+    auto sample_shape = in_shape.tensor_shape_span(i);
+    if (volume(sample_shape) == 0) {
+      out[i] = 0;  // degenerate sample - no reduction possible
+      continue;
+    }
```
Comment (szalpal) [inline dali/kernels/reduce/reduce_setup_utils.h:182]: Hmmm, I'm lost in the 0-dim specialization. Isn't, in case of 0-dim, reduction also impossible, but `volume==1`?

Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:182]: It's not 0-dim. It's 0-volume, regardless of number of dimensions. Still, thanks for getting my attention to this, because it's actually a bug - it should be `max<decltype(Out[0])>` not zero - the reduction is infinite.

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 3 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:142]: I think we already throw other exceptions in kernels (but not in operators).

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 2 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:139]: Actually, I have more stringent limitations in other places, depending on the kind of preprocessing used. I'll document them and add appropriate validation.

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 4 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:112]: It's what we use in TensorShape.

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+constexpr enable_if_t<std::is_integral<T>::value, int> ilog2(T x) {
+  int n = 0;
+  while (x >>= 1)
+    n++;
+  return n;
+}
+
+
+template <typename OutType = uint64_t, typename BitIndices>
+OutType to_bit_mask(const BitIndices &bit_indices) {
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:42]: Yes, it will output incorrect mask.
I though about moving this to core/util.h - in that case the thing I'd expect from such a function is _speed_. For the sake of if, I'll not add verification here - especially if I mark this function as `__host__ __device__` - I've had quite painful encounter with `assert` in device code eating up inordinate number of registers.

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+namespace kernels {
+namespace reduce_impl {
+
+template <typename T>
+constexpr enable_if_t<std::is_integral<T>::value, int> ilog2(T x) {
+  int n = 0;
+  while (x >>= 1)
+    n++;
+  return n;
+}
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:38]: Kidding, yes?

File: dali/kernels/reduce/reduce_gpu.h [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+namespace kernels {
+
+template <typename Out, typename In>
+class SumGPU {
+ public:
+  KernelRequirements Setup(KernelContext &ctx,
+                           const TensorListShape<> &in_shape,
+                           span<const int> axes);
+
+  void Run(KernelContext &ctx, const OutListGPU<Out> &out, const InListGPU<In> &in);
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu.h:32]: Yes.... I'll remove that sentence from there... in a separate PR ;)

File: dali/kernels/reduce/reduce_gpu.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+#include <memory>
+#include "dali/kernels/kernel.h"
+#include "dali/core/tensor_view.h"
+
+namespace dali {
+namespace kernels {
+
+template <typename Out, typename In>
+class SumGPU {
+ public:
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu.h:27]: Oh, this file is a stub, really. There will be more reduction types here.

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 4 above)
Comment (szalpal) [reply dali/kernels/reduce/reduce_setup_utils.h:112]: How about this then?
```
  SmallVector<int, MAX_NDIMS> non_reduced;
  SmallVector<int, 6 /* max number of dims */> non_reduced;
```

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 6 above)
Comment (szalpal) [reply dali/kernels/reduce/reduce_setup_utils.h:182]: I know it's 0-volume. But when we have 0-dim (a scalar), reduction is also impossible, isn't it? So I suppose this condition should look something like
```
if (volume(sample_shape) == 0 || sample_shape.ndims() == 0)
```

Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:215]: It's not like it. When reducing, you force a reduced dimension to 1. If the tensor is empty (0 volume) you can't conjure up the data to be reduced.
A scalar, on the other hand, can't be reduced due to a totally different condition - there's no valid axis index.
I wonder what I should do with these 0-sized tensors - when not all axes are reduced and the output is produced with `keep_dims=true`, I can still set the reduced extents to 1 and non-reduced ones to 0, so the output is empty.

**Comment by mzient** [comment]
!build

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (szalpal) [reply dali/kernels/reduce/reduce_setup_utils.h:49]: I just wanted to point out, that it's not obvious and the incorrect mask will be an error that's hard to spot later on. How much speed would we sacrifice by verifying value here? In per-cents?

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+  int d = in_shape.sample_dim();
+
+  uint64_t mask = to_bit_mask(axes);
+
+  int i;
+
+  // skip leading degenerate dimensions
+  for (i = 0; i < d && is_degenerate_dim(in_shape, i); i++) {}
+
+  if (i < d) {
```
Comment (klecki) [inline dali/kernels/reduce/reduce_setup_utils.h:86]: Can you maybe reverse the condition, do early exit in if case, and do the rest without else?

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 12]
```diff
@@ -0,0 +1,228 @@
...
+    n++;
+  return n;
+}
+
+/**
+ * @brief Returns an integer where bits at indicies in `bit_indices` are set to 1.
+ * @remarks Indices that are outside the bit-width of OutType are ignored.
+ */
+template <typename OutType = uint64_t, typename BitIndices>
+OutType to_bit_mask(const BitIndices &bit_indices) {
```
Comment (klecki) [inline dali/kernels/reduce/reduce_setup_utils.h:49]: I also prefer snake_case, but shouldn't this be `ToBitMask`?

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 13]
```diff
@@ -0,0 +1,1092 @@
...
+      }
+    } else {
+      buf_sizes.AddIO<Out>(stage.output_elements());
+    }
+  }
+
+  /**
+   * @brief Initializes processing where no per-sample dimension is reduced
+   *
+   * The reduction is either pre- and post-processing only or a of repsective elements
```
Comment (klecki) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:428]: respective

Comment (klecki) [reply dali/kernels/reduce/reduce_gpu_impl.cuh:428]: also what's up with: `or a of ...`

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 14]
```diff
@@ -0,0 +1,1092 @@
...
+  Sample,  ///< Reduce samples separately
+  Block,   ///< Reduce contiguous samples separately
+  Inner,   ///< Reduce inner dimension
+  Middle,  ///< Reduce middle or outer dimension
+  None,    ///< No reduction - can happen when reduced dimension is degenerate
+  Fold,    ///< Reduce samples, keep spatial extents
+};
+
+struct ReductionShape {
+  int64_t outer, inner, reduced_in, reduced_out;
```
Comment (klecki) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:49]: TBH, I have no idea what those variables are supposed to describe.

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 15]
```diff
@@ -0,0 +1,1092 @@
...
+#include "dali/core/traits.h"
+
+namespace dali {
+namespace kernels {
+namespace reduce_impl {
+
+enum class ReductionKind {
+  All,     ///< Reduce contiguous memory
+  Sample,  ///< Reduce samples separately
+  Block,   ///< Reduce contiguous samples separately
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:41]: So we assume that samples can be not continuous?

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 16]
```diff
@@ -0,0 +1,1092 @@
...
+    input_offsets.resize(n_in);
+    for (int i = 0; i < n_in; i++) {
+      input_offsets[i] = ofs;
+      ofs += shape[i].input_elements();
+    }
+
+    // if the stage produces contiguous output or just one sample, we only
+    // need one output pointer - otherwise we need a pointer for each sample separately
+    int n_out = kind == ReductionKind::All ||     // only one sample in output
+                kind == ReductionKind::Block ||   // contiguous output
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:96]: Nitpick, you can keep the same order as this values appears in the enum.

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 17]
```diff
@@ -0,0 +1,1092 @@
...
+  vector<ReductionShape> shape;
+  vector<int64_t> input_offsets, output_offsets;
+};
+
+
+struct TempBufferSizes {
+  int64_t io_buffers = 0;
+  int64_t param_buffers = 0;
+
+  void Reserve(TempBufferSizes sz) {
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:118]: It doesn't reserve anything, I don't know if this is a fortunate name.

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 18]
```diff
@@ -0,0 +1,1092 @@
...
+  KernelRequirements Setup(KernelContext &ctx,
+                           const TensorListShape<> &in_shape,
+                           span<const int> axes,
+                           bool keep_dims,
+                           bool reduce_batch) {
+    if (in_shape.sample_dim() > 64)
+      throw std::range_error("Reduce supports up to 64 dimensions");
+    reduce_batch_ = reduce_batch;
+    CheckAxes(axes, in_shape.sample_dim());
+    if (reduce_batch)
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:224]:
```suggestion
    if (reduce_batch_)
```

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 19]
```diff
@@ -0,0 +1,1092 @@
...
+
+  /// DO NOT redefine in CRTP derived class
+  identity GetPostprocessor(std::false_type is_last) const {
+    return {};
+  }
+
+  void CalculateTempBuffers(ScratchpadEstimator &se) {
+    int nstages = stages_.size();
+    buffer_sizes_ = {};
+    for (int stage_idx = 0; stage_idx < nstages; stage_idx++) {
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:327]:
```suggestion
    for (auto& stage: stages_) {
```

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 20]
```diff
@@ -0,0 +1,1092 @@
...
+      auto &stage = stages_[stage_idx];
+      TempBufferSizes stage_buffers;
+      CalculateTempBuffers(stage_buffers, stage);
+      buffer_sizes_.Reserve(stage_buffers);
+    }
+
+    // now reserve the scratchpad - overaligned to 64 bytes, just in case
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:334]: In case of what? It seem like a war for some access beyond the array.

File: dali/kernels/reduce/reduce_gpu_impl_test.cu [hunk 21]
```diff
@@ -0,0 +1,657 @@
...
+#include "dali/kernels/scratch.h"
+#include "dali/test/test_tensors.h"
+#include "dali/test/tensor_test_utils.h"
+#include "dali/core/tensor_shape_print.h"
+
+namespace dali {
+
+
+template <typename T, int ndim>
+std::ostream &operator<<(std::ostream &os, const TensorView<StorageCPU, T, ndim> &t) {
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_gpu_impl_test.cu:28]: Please move to come common utils.

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 22]
```diff
@@ -0,0 +1,1092 @@
...
+    stage.index = 0;
+    stage.is_last = true;
+    stage.kind = reduce_batch_ ? ReductionKind::Fold : ReductionKind::None;
+    int N = in_shape_.num_samples();
+    stage.shape.resize(N);
+    for (int i = 0; i < N; i++) {
+      ReductionShape &rs = stage.shape[i];
+      rs.inner = volume(in_shape_.tensor_shape_span(i));
+      rs.outer = rs.reduced_in = 1;
+      rs.reduced_out = (i == 0 || !reduce_batch_) ? 1 : 0;
```
Comment (klecki) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:444]: What's with reduced_in?

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 23]
```diff
@@ -0,0 +1,1092 @@
...
+      ReductionStage &stage = stages_[s];
+      if (stage.kind == ReductionKind::All) {
+        // ReduceAll cannot happen at stage 0
+        assert(s > 0 || nsamples == 1);
+        int64_t r = 0;
+        if (s > 0) {
+          for (auto &rs : stages_[s-1].shape)
+            r += rs.reduced_out;
+        } else {
+          r = max_reduced_extent;
```
Comment (klecki) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:529]: if ReduceAll cannot happen at `s == 0`, and we even assert this, the whole `if (s > 0) {} else {}` checks are superfluous, also the `r = max_reduced_extent` seems a bit weird.

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:49]: If I put an assert there and it makes me change CUDA block size? Perhaps 50%.

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 24]
```diff
@@ -0,0 +1,1058 @@
...
+        r = CalcReducedExtent(r, remaining_stages);
+        stage.shape[0].reduced_out = r;
+      } else {
+        int64_t max_in = 0;
+        if (s > 0) {
+          for (int i = 0; i < nsamples;  i++) {
+            auto &rs = stage.shape[i];
+            int64_t r = stages_[s-1].shape[i].reduced_out;
+            rs.reduced_in = r;
+            rs.inner = rs.outer = 1;
```
Comment (klecki) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:549]: Why the reduced_in, inner and outer are set only in the `if (s > 0)` case?

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:49]: Anyway, I think you're looking at it as a high level function. It's a _very low level_ function. It's like a glorified bit shift operator - it certainly doesn't throw when you do 1<<100.

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 25]
```diff
@@ -0,0 +1,194 @@
...
+    n++;
+  return n;
+}
+
+
+template <typename OutType = uint64_t, typename BitIndices>
+OutType to_bit_mask(const BitIndices &bit_indices) {
+  OutType mask = 0;
+  for (int idx : bit_indices)
+    mask |= OutType(1) << idx;
```
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:52]: It has a big potential for generating stupid warnings, that's certain. As for the "improvement" - I fail to see any.

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 12 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:49]: I'll move to_bit_mask and ilog2 to code/util.h - just not now, because it would cause massive rebuilds when switching branches.

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 26]
```diff
@@ -0,0 +1,1058 @@
...
+
+    for (int s = 0; s < substages; s++) {
+      ReductionStage &stage = stages_[s];
+      stage.index = s;
+      // stage 0 is never ReduceAll, because that would require contiguous input
+      stage.kind = s == 0 ? ReductionKind::Sample
+                          : reduce_batch_ ? ReductionKind::All : ReductionKind::Block;
+      stage.shape.resize(stage.kind == ReductionKind::All ? 1 : nsamples);
+    }
+
```
Comment (klecki) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:517]: Correct me if I'm wrong, but is the whole processing below just setting up progressively smaller stages with condition stage[i].reduced_in = stage[i-1].reduced_out?
And inner = outer = 1 for all stages?
Isn't the reduce_batch_ step happening just ather the first one, and than you should be able to handle this with two variants of simple loop for stages 1..substages-1, in one you have `nsmaples` shapes in other only 1.

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 27]
```diff
@@ -0,0 +1,1058 @@
...
+  void Run(KernelContext &kctx, const OutListGPU<Out> &out, const InListGPU<In> &in) {
+    assert(!stages_.empty());
+    Context ctx;
+    auto host_mem_size = buffer_sizes_.param_buffers;
+    auto gpu_mem_size = buffer_sizes_.param_buffers + buffer_sizes_.io_buffers;
+
+    ctx.stream = kctx.gpu.stream;
+    ctx.work_area.buffer_sizes = buffer_sizes_;
+    ctx.work_area.host_memory = kctx.scratchpad->Allocate<char>(AllocType::Host, host_mem_size, 64);
+    ctx.work_area.gpu_memory = kctx.scratchpad->Allocate<char>(AllocType::GPU, gpu_mem_size, 64);
```
Comment (klecki) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:998]: You're not using the same alloc&pad approach to gpu param & io buffers (in setup you configure them in two separate steps). It probably won't matter, but that's probably the only issue I will be able to notice, as the CalculateTempBuffers is incomprehensible for me rn (at least till I understand the memory use of the whole kernel).

File: dali/kernels/reduce/reduce_gpu_impl.cuh [hunk 28]
```diff
@@ -0,0 +1,1058 @@
...
+    return samples;
+  }
+
+  template <typename StageOut, typename StageIn, bool is_first, bool is_last>
+  void LaunchStage(Context &ctx, const ReductionStage &stage,
+                   ReductionKindTag<ReductionKind::All>) {
+    assert(!is_first || ctx.input.is_contiguous());
+    WorkArea &wa = ctx.work_area;
+
+    auto pre = This().GetPreprocessor(bool_const<is_first>());
```
Comment (klecki) [inline dali/kernels/reduce/reduce_gpu_impl.cuh:777]: Can the `All` reduction be first?

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 11 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:86]: Why not.

File: dali/kernels/reduce/reduce_gpu_impl.cuh (hunk 15 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu_impl.cuh:41]: Yes.

File: dali/kernels/reduce/reduce_gpu_impl.cuh (hunk 18 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu_impl.cuh:224]: I can do it, but using the parameter is safer, in case the check is moved above the member assignment.

File: dali/kernels/reduce/reduce_gpu_impl.cuh (hunk 22 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu_impl.cuh:451]: It's above.

File: dali/kernels/reduce/reduce_gpu_impl.cuh (hunk 26 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu_impl.cuh:517]: I don't see how this would be simpler.

File: dali/kernels/reduce/reduce_gpu_impl.cuh (hunk 23 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu_impl.cuh:536]: Stale comment - look at the assert again.

File: dali/kernels/reduce/reduce_gpu_impl.cuh (hunk 24 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu_impl.cuh:549]: It's missing, but they're not used in ReduceAll/ReduceBlock anyway.

File: dali/kernels/reduce/reduce_gpu_impl.cuh (hunk 28 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu_impl.cuh:777]: If there's just one sample in the input.

File: dali/kernels/reduce/reduce_gpu_impl.cuh (hunk 20 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu_impl.cuh:334]: The calculation of buffer size happens elsewhere and here we don't know what types are allocated - so we're very conservative about alignment.

File: dali/kernels/reduce/reduce_gpu_impl.cuh (hunk 27 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu_impl.cuh:998]: Right you are. Actually, I should align param_buffer's size and allocate - and reserve - all gpu memory in one go.

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

**Comment by mzient** [comment]
!build

File: dali/kernels/reduce/reduce_gpu.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu.h:27]: OK, now it's not a stub, but it's still partial. Mean, RMS and StdDev will come in a followup.

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 29]
```diff
@@ -0,0 +1,229 @@
...
+ *
+ * @param in_shape   shape of the input
+ * @param axes       list of reduced axes; axis indices must be <= 63
+ *
+ * A dimension can be collapsed with its neighbor if:
+ *  - its extent is 1 in all samples
+ *  - both dimensions are reduced or non-reduced.
+ */
+template <typename Axes, typename DimGroups, int ndim>
+void SimplifyReduction(Axes &out_axes, DimGroups &dim_groups,
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_setup_utils.h:73]:
```suggestion
void SimplifyReduction(Axes &out_axes, DimGroups &out_dim_groups,
```
?

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 30]
```diff
@@ -0,0 +1,229 @@
...
+    if (!(mask & (1 << a)))
+      non_reduced.push_back(a);
+  }
+
+  auto first_sample_shape = tls.tensor_shape_span(0);
+
+  for (int i = 1; i < tls.num_samples(); i++) {
+    auto sample_shape = tls.tensor_shape_span(i);
+    for (int a : non_reduced) {
+      DALI_ENFORCE(sample_shape[a] == first_sample_shape[a], make_string(
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_setup_utils.h:149]: DO you want to use `DALI_ENFORCE` inside kernel? I think the plain exception is the way.

File: dali/kernels/reduce/reduce_setup_utils.h [hunk 31]
```diff
@@ -0,0 +1,229 @@
...
+      throw std::out_of_range(make_string("Axis index out of range: ", a, " not in 0..", ndim-1));
+    uint64_t amask = 1ul << a;
+    if (mask & amask)
+      throw std::invalid_argument(make_string("Duplicate axis index ", a));
+    mask |= amask;
+  }
+}
+
+
+inline void CalculateReducedShape(TensorListShape<> &out_shape,
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_setup_utils.h:178]: Any doxygen doc, please.

File: dali/kernels/reduce/reduce_gpu_impl.cuh (hunk 20 above)
Comment (JanuszL) [reply dali/kernels/reduce/reduce_gpu_impl.cuh:383]: So I would rephrase the comment to explain that.

File: dali/kernels/reduce/reduce_gpu_impl_test.cu [hunk 32]
```diff
@@ -0,0 +1,533 @@
...
+
+  std::mt19937_64 rng(12345);
+
+  TestTensorList<uint8_t> in;
+  in.reshape(in_shape);
+  auto in_cpu = in.cpu();
+  UniformRandomFill(in_cpu, rng, 0, 255);
+
+  TestTensorList<int64_t> out;
+  out.reshape(req.output_shapes[0]);
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_gpu_impl_test.cu:373]: Any particular reason why you don't call `cudaMemset` lik in the previous test case?

File: dali/kernels/reduce/reduce_gpu_test.cc [hunk 33]
```diff
@@ -0,0 +1,87 @@
...
+  TestTensorList<uint8_t> in;
+  in.reshape(in_shape);
+  auto in_cpu = in.cpu();
+  UniformRandomFill(in_cpu, rng, 0, 255);
+
+  TestTensorList<uint64_t> out;
+  out.reshape(req.output_shapes[0]);
+  sum.Run(ctx, out.gpu(), in.gpu());
+
+  auto out_cpu = out.cpu(ctx.gpu.stream);
```
Comment (JanuszL) [inline dali/kernels/reduce/reduce_gpu_test.cc:59]: Doesn't it call `cudaMemcpyAsync`? It requires to sync to have the memory available on the CPU, right?

Comment (mzient) [reply dali/kernels/reduce/reduce_gpu_test.cc:59]: AFAIK copying to non-pinned memory is implicitly synchronized with the host.

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 29 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:73]: will do

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 30 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:149]: I'll replace it with appropriate error (invalid argument or logic error).

File: dali/kernels/reduce/reduce_gpu_impl.cuh (hunk 27 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu_impl.cuh:1075]: Done (param_buffers are aligned in CalculateTempBuffers.

File: dali/kernels/reduce/reduce_setup_utils.h (hunk 31 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_setup_utils.h:172]: Done.

File: dali/kernels/reduce/reduce_gpu_impl_test.cu (hunk 32 above)
Comment (mzient) [reply dali/kernels/reduce/reduce_gpu_impl_test.cu:373]: I only used cudaMemset in case of degenerate input (zero size) - in that cases, the expected value is zero, which may coincide with how the memory is initialized on allocation. Here, I have some real value to compare against and it's exceedingly unlikely that the memory will happen to contain the expected result.

**Comment by mzient** [comment]
!build

File: dali/kernels/reduce/reduce_gpu_test.cc (hunk 33 above)
Comment (JanuszL) [reply dali/kernels/reduce/reduce_gpu_test.cc:59]: OK

File: dali/kernels/reduce/reduce_gpu_impl_test.cu (hunk 32 above)
Comment (JanuszL) [reply dali/kernels/reduce/reduce_gpu_impl_test.cu:373]: OK

**Comment by mzient** [comment]
!build
